### ISSUE TITLE
Port NSS for Haiku

### ISSUE THREAD

**Issue by kenz-gelsoft** [kenz-gelsoft/gecko-dev#16]
As BeOS support was removed between esr102 and esr128.

https://phabricator.services.mozilla.com/D37215

I tried `patch -p1 -R < D37215.1723043856.patch` but it doesn't apply cleanly.

Fix those or port NSS to new platform as a UNIX variant.

**Comment by kenz-gelsoft** [comment]
there is no fdopen() define in esr128. I assume Haiku has working fdopen().

zutil.h.rej
```diff
--- lib/zlib/zutil.h
+++ lib/zlib/zutil.h
@@ -168,11 +168,11 @@
 
 #ifdef __APPLE__
 #  define OS_CODE 19
 #endif
 
-#if defined(RISCOS)
+#if defined(_BEOS_) || defined(RISCOS)
 #  define fdopen(fd,mode) NULL /* No fdopen() */
 #endif
 
 #if (defined(_MSC_VER) && (_MSC_VER > 600)) && !defined __INTERIX
 #  if defined(_WIN32_WCE)
```

**Comment by kenz-gelsoft** [comment]
Patching haiku's patcheset on it
```
~/src/gecko-dev/security/nss> patch -p1 < ../../nss.patch 
patching file coreconf/BeOS.mk
patching file coreconf/arch.mk
Hunk #1 succeeded at 34 with fuzz 2 (offset 1 line).
patching file lib/dbm/include/mcom_db.h
patching file lib/freebl/stubs.c
Hunk #1 succeeded at 813 (offset 95 lines).
patching file cmd/shlibsign/Makefile
patching file coreconf/BeOS.mk
patching file coreconf/BeOS.mk
patching file coreconf/BeOS.mk
patching file lib/freebl/Makefile
Hunk #1 succeeded at 366 (offset 33 lines).
patching file coreconf/BeOS.mk
patching file coreconf/BeOS.mk
patching file cmd/httpserv/httpserv.c
Hunk #1 succeeded at 1321 (offset -4 lines).
patching file cmd/selfserv/selfserv.c
Hunk #1 succeeded at 2881 (offset 213 lines).
patching file coreconf/BeOS.mk
```

**Comment by kenz-gelsoft** [comment]
Trying to build on https://github.com/kenz-gelsoft/gecko-dev/pull/17

**Comment by kenz-gelsoft** [comment]
I think we can port NSS(or NSPR) differently as Haiku has higher complience to POSIX than original BeOS.

Can be cleanly ported again as a UNIX variant. but for now, I chose proven route.

**Comment by kenz-gelsoft** [comment]
randomly fails creating (build) dependencies. but retrying seems to work.
```
 1:05.27    Compiling firefox-on-glean v0.1.0 (/boot/home/src/gecko-dev/toolkit/components/glean/api)
 1:06.69 clang-12: warning: argument unused during compilation: '-fstack-clash-protection' [-Wunused-command-line-argument]
 1:13.59 dom/file/uri
 1:13.68 clang-12: warning: argument unused during compilation: '-fstack-clash-protection' [-Wunused-command-line-argument]
 1:13.93 clang-12: warning: argument unused during compilation: '-fstack-clash-protection' [-Wunused-command-line-argument]
 1:24.56    Compiling gecko-profiler v0.1.0 (/boot/home/src/gecko-dev/tools/profiler/rust-api)
 1:25.85 warning: hard linking files in the incremental compilation cache failed. copying files instead. consider moving the cache directory to a file system which supports hard linking in session dir `/boot/home/src/gecko-dev/obj-ff-dbg/x86_64-unknown-haiku/debug/incremental/gecko_profiler-310wm4m3uautd/s-gyqttk3701-1s0kctt-working`
 1:26.87 warning: `gecko-profiler` (lib) generated 1 warning
 1:26.87    Compiling mozglue-static v0.1.0 (/boot/home/src/gecko-dev/mozglue/static/rust)
 1:27.56 warning: hard linking files in the incremental compilation cache failed. copying files instead. consider moving the cache directory to a file system which supports hard linking in session dir `/boot/home/src/gecko-dev/obj-ff-dbg/debug/incremental/build_script_build-2ynfomk73rh5k/s-gyqttlgom7-193imq4-working`
 1:29.59 warning: `mozglue-static` (build script) generated 1 warning
 1:30.39 dom/filesystem/compat
 1:30.49 clang-12: warning: argument unused during compilation: '-fstack-clash-protection' [-Wunused-command-line-argument]
 1:34.64 warning: hard linking files in the incremental compilation cache failed. copying files instead. consider moving the cache directory to a file system which supports hard linking in session dir `/boot/home/src/gecko-dev/obj-ff-dbg/x86_64-unknown-haiku/debug/incremental/firefox_on_glean-jppux9wxyf55/s-gyqtt8nsah-eok4g5-working`
 1:34.81 error: failed to create dependency graph at `/boot/home/src/gecko-dev/obj-ff-dbg/x86_64-unknown-haiku/debug/incremental/firefox_on_glean-jppux9wxyf55/s-gyqtt8nsah-eok4g5-working/dep-graph.part.bin`: No such file or directory (os error -2147459069)
```

**Comment by kenz-gelsoft** [comment]
neqo-crypto or other security related build errors by old NSS looks resolved.

**Comment by kenz-gelsoft** [comment]
Will be fixed by #17
